### DISCUSSION TITLE
MODE Issue with "Interest" Calculation

### DISCUSSION THREAD

**Discussion by jhicks14** [dtcenter/METplus#1246] [Statistical Computation] [answered]
I'm running the MODE tool on hazard forecasts. Each forecasted hazard shape represents the probability of occurrence; e.g., a value of 0.2 represents 20% probability of a hazard occurring over the given area. Each map has the possibility of containing multiple probability contours, which may be the source of the issue I'm having. Three conv thresholds are specified, so three PostScript images are output. 

When I run the attached METplus config file, the PostScript images look mostly correct, but only one of the three has a valid interest value (see the conv thresh >0.5 image). In this run, the forecast object is equal to 0.6 (60%) and closely resembles the observed object. The interest value is 1 because I set all interest functions to 1 for testing purposes. In the image where the conv thresh >0.3, the forecast object is equal to 0.4 (40%), and the object SHOULD include the area occupied by the observed object. However, the interest value shows up as "NA". I have the same issue with the final image (conv thresh >0.0), where the forecast object is equal to 0.2 (20%). Is there any way I can get the area occupied by the lower probabilities to include the higher probability areas (where there are observed objects)? In other words, the observed object is technically within the forecasted object for the 20% and 40% contours, but no interest value is output. In reality, the interest values should increase (in this particular case) with each increasing conv threshold.

![mode_cold_hazard_000000L_20180203_000000V_000000A_R1_T1-0](https://user-images.githubusercontent.com/90361476/140851696-bd63fc90-3615-47d6-bfc2-d6ac56f89c14.png)

![mode_cold_hazard_000000L_20180203_000000V_000000A_R2_T2-0](https://user-images.githubusercontent.com/90361476/140851707-a131eafb-e5e9-4c81-9fe1-d9710e85fccb.png)

![mode_cold_hazard_000000L_20180203_000000V_000000A_R3_T3-0](https://user-images.githubusercontent.com/90361476/140851396-152d475e-c157-4563-bfee-f1b1a1b2f31e.png)

[MODE_hazards.txt](https://github.com/dtcenter/METplus/files/7501477/MODE_hazards.txt)

**Comment by jprestop** [comment, marked as answer]
HI @jhicks14.

It sounds likely that the reason for no interest value being computed is the "max_centroid_dist" MODE config option. Increasing that to a much larger value should enable interest pairs to be computed. The units are grid squares so you could just set it to Nx + Ny (where those are the number of grid points in the x and y directions). No two grid points would be further apart than that.  Perhaps we should support it being set to "NA" in which case we wouldn't check it at all? What are your thoughts on that?

At verbosity level 5, MODE prints much more detailed info about each object comparison. For example:
```
DEBUG 5: Computing total interest for single pair F007_O004.
DEBUG 5: 
DEBUG 5: Total Interest = -9999.000
DEBUG 5: Centroid Distance (28.762) > Max Centroid Distance (20.000)
```

Here is the [documentation](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=max_centroid_dist#max-centroid-dist) for the max_centroid_dist config option in the MET User’s Guide:

> The “max_centroid_dist” entry specifies the maximum allowable distance in grid squares between the centroids of objects for them to be compared. Setting this to a reasonable value speeds up the runtime enabling MODE to skip unreasonable object comparisons.
> 
> `max_centroid_dist = 800.0/grid_res;`

And the documentation from the [MODE section](https://met.readthedocs.io/en/latest/Users_Guide/mode.html?highlight=max_centroid) of the User’s Guide:

> `max_centroid_dist = 800/grid_res;`
> 
> Computing the attributes for all possible pairs of objects can take some time depending on the numbers of objects. The max_centroid_dist entry is used to specify how far apart objects should be in order to conclude that they have no chance of matching. No pairwise attributes are computed for pairs of objects whose centroids are farther away than this distance, defined in terms of grid units. Setting this entry to a reasonable value will improve the execution time of the MODE tool. By default, the maximum centroid distance is defined in terms of the previously defined grid_res entry.

If you are using METplus, here is an [example](https://metplus.readthedocs.io/en/latest/generated/met_tool_wrapper/MODE/MODE.html?highlight=max_centroid_dist) of how you could set the max_centroid_dist in your METplus configuration file:

> //
> // Maximum centroid distance for objects to be compared
> //
> //max_centroid_dist =
> ${METPLUS_MAX_CENTROID_DIST}

For more information on how METplus controls the MET configuration variables we have this [documentation] (https://metplus.readthedocs.io/en/latest/Users_Guide/systemconfiguration.html#metplus-control-met), but don’t hesitate to reach out with any questions about how to configure METplus/MET.

**Reply by jhicks14** [reply]
The change to the **max_centroid_dist** is exactly what I was looking for. Thanks again for all of you help!

I think having the option to set the max centroid distance to "NA" could be helpful for testing purposes rather than setting to an unrealistic distance. That could be helpful especially for those without much experience using the MODE tool.

**Reply by jprestop** [reply]
Hi @jhicks14.  Great!  I'm glad to hear that is what you were looking for.  Thank you also for your feedback regarding the ability to set the max centroid distance to "NA".  @JohnHalleyGotway has created a [GitHub Issue](https://github.com/dtcenter/MET/issues/1964) for that enhancement.  

I'd like to mark this discussion as being answered and "lock it" to prevent future posts. That's how we encourage users to create new discussions for new questions. But, I wanted to give you an opportunity to comment on it before doing so. Please feel free to select the answer to your question in this Discussion. That'll help future users with similar questions find answers.
